PubSub/src/MessageBuilder.php

```diff
@@ -0,0 +1,179 @@
+<?php
+/**
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\PubSub;
+
+use Google\Cloud\Core\ArrayTrait;
+
+/**
+ * Builds a PubSub Message.
+ *
+ * This class may be used to build an API-compliant message for publication.
+ *
+ * This class is immutable and each build method returns a new instance with
+ * your changes applied.
+ *
+ * Example:
+ * ```
+ * use Google\Cloud\PubSub\MessageBuilder;
+ * use Google\Cloud\PubSub\PubSubClient;
+ *
+ * $client = new PubSubClient();
+ * $topic = $client->topic($topicId);
+ *
+ * $builder = new MessageBuilder();
+ * $builder = $builder->setData('hello friend!')
+ *     ->addAttribute('from', 'Bob')
+ *     ->addAttribute('to', 'Jane');
+ *
+ * $topic->publish($builder->build());
+ * ```
+ */
+class MessageBuilder
+{
+    use ArrayTrait;
+
+    /**
+     * @var array
+     */
+    private $message;
+
+    /**
+     * @param array $message The initial message data.
+     */
+    public function __construct(array $message = [])
+    {
+        $this->message = $message;
+    }
+
+    /**
+     * Set the message data.
+     *
+     * Do not base64-encode the value; If it must be encoded, the client will
+     * do it on your behalf.
+     *
+     * Example:
+     * ```
+     * $builder = $builder->setData('Hello friend!');
+     * ```
+     *
+     * @param string $data The message data field. If this field is empty, the
```

Comment:
```suggestion
     * @param string|null $data The message data field. If this field is empty, the
```

Given the note about this being potentially empty, WDYT?

Reply:
I was trying to convey that you cannot build a message without data or at least one attribute. This method shouldn't accept null. I added a note to the class documentation instead and removed this.